### PR TITLE
feat:[CDS-81282]: cd stage summary resource for notifications

### PR DESCRIPTION
**Pull request by naman-goenka**
## Describe your changes
- Contract for getting cd stage summary.
- For upcoming and currently executing stages, plan creation summary contract filtered via stage identifiers and plan execution and scope. Result can have expressions as well. 
- For already executed stages, execution summary contract filtered via scope and stage execution identifiers.
- Index for both the queries was validated manually.
- Projection is used in stage execution info based query as the entity has lot of other irrelevant info in this context
- Spec: [Phase 1](https://harness.atlassian.net/wiki/spaces/CDNG/pages/21520647005/Enhance+Approval+notifications+to+add+details+related+to+services+environment+artifacts+CG+Parity#How-do-we-plan-to-add-more-details-in-the-notifications%5BhardBreak%5D%5BhardBreak%5DPhase-1)
- Testing - [Dev test plan](https://docs.google.com/spreadsheets/d/1pfjy2dhRo5l-qiL94Fv_MBjs3623FqlCIKTdWWL7F2Y/edit#gid=0)
- **Stage Execution Summary**
```curl --location 'https://localhost:8181/ng/api/cdStageSummary/listStageExecutionFormattedSummary?accountIdentifier=kmpySmUISimoRrJL6NL73w&orgIdentifier=default&projectIdentifier=defaultProject&stageExecutionIdentifiers=Cvwgj4c-Q3irukivAIIz-A&stageExecutionIdentifiers=Q8VdanYESCKlGf1dctgWnw&stageExecutionIdentifiers=X3P47UaERmWaF1XbBd8SFw&stageExecutionIdentifiers=htFlEY9XSs2tcvrEQC_QhQ&stageExecutionIdentifiers=gbfvdcx' \```

Screenshot 2023-11-08 at 9 01 09 PM

- **Plan creation Summary**
```curl --location 'https://localhost:8181/ng/api/cdStageSummary/listStagePlanCreationFormattedSummary?accountIdentifier=kmpySmUISimoRrJL6NL73w&orgIdentifier=default&projectIdentifier=defaultProject&stageIdentifiers=test&planExecutionId=bpKK7LD1Slez6VwJG8i8fw&stageIdentifiers=test1&stageIdentifiers=test2&stageIdentifiers=test3' \```

Screenshot 2023-11-08 at 9 01 56 PM

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55095)

### CODE REVIEW

**Comment by naman-goenka** [comment]
trigger uts

**Comment by naman-goenka** [comment]
trigger uts

**Comment by naman-goenka** [comment]
trigger smartchecks

**Comment by naman-goenka** [comment]
trigger uts

**Comment by naman-goenka** [comment]
trigger smartchecks

**Comment by naman-goenka** [comment]
unrelated UTs are failing

**Review by jennyjames21 — APPROVED** [review]

File: 970-ng-commons/src/main/java/io/harness/cd/CDStageSummaryConstants.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import io.harness.annotations.dev.OwnedBy;
+import io.harness.annotations.dev.ProductModule;
+
+import lombok.experimental.UtilityClass;
+
+@CodePulse(module = ProductModule.CDS, unitCoverageRequired = false,
+    components = {HarnessModuleComponent.CDS_SERVICE_ENVIRONMENT})
+@OwnedBy(CDC)
+@UtilityClass
+public class CDStageSummaryConstants {
```
Comment (abhinavmittal-harness) [inline 970-ng-commons/src/main/java/io/harness/cd/CDStageSummaryConstants.java:23]: nit: the ideal way to define a constants class is generally making the class final and have a private constructor to hide default public one

**Review by abhinavmittal-harness — APPROVED** [review]

**Review by archit-harness — COMMENTED** [review]
> Reviewed 1 of 17 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55095)* status: 27 file(s) to review.

**Comment by abhinavmittal-harness** [comment]
triger review

**Comment by abhinavmittal-harness** [comment]
trigger review

File: 970-ng-commons/src/main/java/io/harness/cd/CDStageSummaryConstants.java (hunk 1 above)
Comment (naman-goenka) [reply 970-ng-commons/src/main/java/io/harness/cd/CDStageSummaryConstants.java:23]: done

**Comment by naman-goenka** [comment]
trigger uts

**Comment by naman-goenka** [comment]
trigger smartchecks

**Comment by naman-goenka** [comment]
trigger smartchecks, uts

**Comment by naman-goenka** [comment]
trigger messagecheck

**Review by raghuAtWings — APPROVED** [review]

**Review by abhinavmittal-harness — APPROVED** [review]

**Comment by naman-goenka** [comment]
trigger review

**Comment by naman-goenka** [comment]
triaged failing tests to be unrelated

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/SingleServiceEnvDeploymentStageDetailsInfo.java [hunk 2]
```diff
@@ -38,4 +40,22 @@ public class SingleServiceEnvDeploymentStageDetailsInfo implements DeploymentSta
   @Nullable private String infraName;
   @Nullable private String serviceIdentifier;
   @Nullable private String serviceName;
+  public static final String NOT_AVAILABLE = "NA";
+
+  @Override
+  public CDStageSummaryResponseDTO getFormattedStageSummary() {
+    String environment = StringUtils.isBlank(envName) ? envIdentifier : envName;
```
Comment (puthrayaharness) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/SingleServiceEnvDeploymentStageDetailsInfo.java:47]: You can use `StringUtils.defaultIfBalnk` which will be a lot more cleaner

**Review by puthrayaharness — APPROVED** [review]

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/SingleServiceEnvDeploymentStageDetailsInfo.java (hunk 2 above)
Comment (naman-goenka) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/creator/plan/stage/SingleServiceEnvDeploymentStageDetailsInfo.java:47]: will do it in a follow up PR

**Pull request merged**